### PR TITLE
[Code coverage] Improve the http module coverage.

### PR DESCRIPTION
**Pull request by knightburton**
- Add new test cases to the existing files.
- Add new test file to cover the request response messages.

Note: The setTimeout function from the outgoungMessage is deleted,
because every parents are implements their own setTimeout function.

IoT.js-DCO-1.0-Signed-off-by: Imre Kiss kissi.szeged@partner.samsung.com

### CODE REVIEW

**Comment by yichoi** [comment]
@knightburton Please rebase it to the master

File: test/run_pass/test_net_httpserver_timeout.js [hunk 1]
```diff
@@ -25,6 +25,8 @@ var server = http.createServer(function(req, res) {
 });
 
 server.listen(3003);
+// Overwrite the default timeout value.
+server.setTimeout(100);
```
Comment (daeyeon) [inline test/run_pass/test_net_httpserver_timeout.js]: There is no verification for this test.

File: src/js/http_outgoing.js [hunk 2]
```diff
@@ -152,8 +152,8 @@ OutgoingMessage.prototype._storeHeader = function(statusLine) {
 
 OutgoingMessage.prototype.setHeader = function(name, value) {
 
-  if (this._headers === null) {
-    this._headers = {};
+  if (util.isNull(this._headers)) {
+    this._headers = [];
```
Comment (daeyeon) [inline src/js/http_outgoing.js]: First of all, this commit is to improve code coverage. So please don't fix any original code. Plus, I don't agree this change. Kindly let me know why using `Array` is needed. Seeing the code below, `this._headers` is still used as `Object` although you set it as `Array`.

File: test/run_pass/test_net_httpserver_timeout.js (hunk 1 above)
Comment (knightburton) [inline test/run_pass/test_net_httpserver_timeout.js]: I've added an assert below this line.

File: src/js/http_outgoing.js (hunk 2 above)
Comment (knightburton) [inline src/js/http_outgoing.js]: You are right! I will separate the coverage and fix in the future. The other hand, you are right again, there is no point to use array instead of object, this was my inattention.

File: test/run_pass/test_net_httpserver_timeout.js (hunk 1 above)
Comment (daeyeon) [inline test/run_pass/test_net_httpserver_timeout.js]: I suggest the test like below.

```
var timer = setTimeout(function () {
    defaultTimeout = true;
}, server.timeout);

server.setTimeout(XXX, function(socket) {
    ....
    assert.equal(defaultTimeout, false);
    clearTimeout(timer);
```

**Comment by yichoi** [comment]
@knightburton please rebase it to the master

File: src/js/http_outgoing.js (hunk 2 above)
Comment (daeyeon) [inline src/js/http_outgoing.js]: #798 is related to this discussion.

**Comment by knightburton** [comment]
@daeyeon I've updated the PR. I would like to know what do you think about it. First of all, I've removed the code fixes (I'll update them in a separate PR later), and isolated every test case into a server-request pair in the request-response test file. In the other hand, I removed the server default timeout overwrite test for now, because I think neither of our solution is not the best. I would like to update this patch with a proper timeout test and invalid utf-8 test later.

File: test/run_pass/test_net_http_request_response.js [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+// the default port is 80 and the default method is GET.
+var request0 = http.get(function(response) {
+  var str = '';
+
+  response.on('data', function(chunk) {
+    str += chunk;
+  });
+
+  response.on('end', function() {
+    assert.equal(str, 'default');
```
Comment (daeyeon) [inline test/run_pass/test_net_http_request_response.js]: This seems too natural connection failure case to test. Do we need to test if our api doesn't work when user doesn't fill up the mandatory field for the api?

File: test/run_pass/test_net_http_request_response.js [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+});
+server4.listen(3008, 5);
+
+var request4 = http.request({
+  method: 'HEAD',
+  port: 3008,
+  path: '/'
+}, function(response) {
+  response.on('end', function() {
+    assert.equal(response.statusCode, 200);
```
Comment (daeyeon) [inline test/run_pass/test_net_http_request_response.js]: Please clearly use `assert` on what you check for each test. Since this compares the status code, it seems to assume that the test needs any data from server. I suggest the below pattern. The same problem is found elsewhere in this code.

```
var isTestCase1Called = false;
var request1 = http.get(function(res) {
  ....
  res.on('end', function () {
      isTestCase1Called = true;
  });
  ....
});


process.on('exit', function () {
  assert.equal(isTestCase1Called, false);
});
``

File: test/run_pass/test_net_http_request_response.js (hunk 3 above)
Comment (knightburton) [inline test/run_pass/test_net_http_request_response.js]: Yes, I think it is necessary. I think we should test the default, natural cases. too.

File: test/run_pass/test_net_http_request_response.js (hunk 4 above)
Comment (knightburton) [inline test/run_pass/test_net_http_request_response.js]: I understand what are you suggest, but also I think this test case is fine because the communication should be OK without any body data and that is why there is the status code check.

File: test/run_pass/test_net_headers.js [hunk 5]
```diff
@@ -29,6 +29,7 @@ var server = http.createServer(function (req, res) {
   });
 
   var endHandler = function () {
+    res.removeHeader('h1'); // this should return.
```
Comment (zherczeg) [inline test/run_pass/test_net_headers.js]: Put comment before this line.

File: test/run_pass/test_net_http_request_response.js (hunk 3 above)
Comment (daeyeon) [inline test/run_pass/test_net_http_request_response.js]: In the case, it seems right to check if the api doesn't crash, or else if it throws any information for the wrong usage. However, assert here checks str although it's even never be called. Confusing.

Comment (daeyeon) [inline test/run_pass/test_net_http_request_response.js]: You maybe misunderstood `options`. It's mandatory, so that the api doesn't need to gurantee the case where user doesn't fill it up. If the given `options` is just an object having nothing, the api fill it up with default values. https://github.com/Samsung/iotjs/blob/master/docs/api/IoT.js-API-HTTP.md#httpgetoptions-callback

Comment (zherczeg) [inline test/run_pass/test_net_http_request_response.js]: Should it throw an error if no options passed then?

Comment (knightburton) [inline test/run_pass/test_net_http_request_response.js]: No, it should not. It is just fills up the options values with default parameters. That is what i've tested, if we can't define any options value, the connection should be fine.

Comment (daeyeon) [inline test/run_pass/test_net_http_request_response.js]: Throwing an error is needed. e.g) calling `server0.listen()` without the first mandatory param, port, throws an error. The criteria is the API reference, not the current observation.

Comment (knightburton) [inline test/run_pass/test_net_http_request_response.js]: Sorry for the late response, then what should I do with this test? Should I remove it for now?

Comment (daeyeon) [inline test/run_pass/test_net_http_request_response.js]: Yes, please remove this test for now. It'd be better to treat it on another commit for changing the condition of `options` to optional.

Comment (knightburton) [inline test/run_pass/test_net_http_request_response.js]: Done, removed the test.

**Comment by daeyeon** [comment]
LGTM
